### ISSUE TITLE
Automatically create collection objects for navigations

### ISSUE THREAD

**Issue by ajcvickers** [dotnet/efcore#451]
When materializing (and maybe elsewhere) collection navigations if the collection navigation is null, then a collection should be created and set automatically.

**Comment by rowanmiller** [comment]
We should make sure we initialize all collections, not just the ones that we need to interact with during materialization.

**Comment by ajcvickers** [comment]
Clearing this for re-triage since the behavior we have may be correct. Currently, if you Include a collection navigation in the query, then the navigation property will be initialized even if there are no related entities. If the navigation property is not included in the query, then the property remains uninitialized. This is the same behavior as EF6. It essentially means that if the navigation has been loaded, then it will not be null. But if it hasn't been loaded, then it may be null. This does not seem unreasonable.

Always initializing all collections could have a significant impact on memory usage when querying for many entities for which the collections do not need to be loaded and will never be used.

However, there is another consideration for EF7: no lazy loading. I think this is really why people are hitting null refs more now in EF7 because it means that accessing an unloaded collection results in a null-ref exception rather than just loading the collection. Related to this, the fixup that happens when two queries are run will not initialize the collection if there are no related entities (as in EF6), which means that this loading workaround coupled with no lazy loading can result in unexpected null-refs.

I think we could consider generating code from the reverse engineering process that puts lazy initialization of collections into the generated classes. (We initialized collections in the EF6 generated code, but we didn't do it lazily.) We could also start doing this in all our code first samples/demos/etc., in the same way as we used to always make the properties virtual to enable lazy loading.
